### PR TITLE
Fix more ICEs in `diagnostic::on_unimplemented`

### PR DESCRIPTION
**Pull request by compiler-errors**
There were 8 other calls to `expect_local` left in `on_unimplemented.rs` -- all of which (afaict) could be turned into ICEs.

I would really like to see validation of `on_unimplemented` separated from parsing, so we only emit errors here:
https://github.com/rust-lang/rust/blob/a60f077c38fe66d05449919842d3d73e3299bbab/compiler/rustc_hir_analysis/src/check/check.rs#L836-L839
...And gracefully fail instead when emitting trait predicate failures, not *ever* even trying to emit an error or a lint. But that's left for a separate PR.

r? @estebank

### CODE REVIEW

**Comment by compiler-errors** [comment]
This should be backported together with #124683.

File: tests/ui/diagnostic_namespace/on_unimplemented_ice.rs [hunk 1]
```diff
@@ -1,17 +0,0 @@
-//@ edition:2021
```
Comment (compiler-errors) [inline tests/ui/diagnostic_namespace/on_unimplemented_ice.rs:1]: Github didn't recognize the file being moved, but I gave it a better name. Also, there's no reason for us to be using `#[test]` here -- it's not necessary to reproduce the ICE; all that is necessary is a foreign crate.

**Comment by apiraino** [comment]
Beta backports approved as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-05-09/near/437830021). A backport PR will be authored by the release team at the end of the current development cycle.

Stable backport: same reasoning as for #124683: does not justify a dot release but this backport could jump on that train, if a dot release is planned for other reasons

@rustbot label +beta-accepted +stable-accepted

**Pull request merged**